<a href="https://colab.research.google.com/github/farihamim11/desktop-tutorial/blob/main/GraphSearch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
sabcg_graph = {
    'S': [('A', 1), ('B', 1)],
    'A': [('C', 1)],
    'B': [('C', 2)],
    'C': [('G', 3)],
    'G': [],
}
h_sabcg = {'S': 2, 'A': 4, 'B': 1, 'C': 1, 'G': 0}

In [ ]:
def astar_priority_sabcg(g,node):
  return g+h_sabcg[node]

In [ ]:
def graph_search(graph, start, goal, priority_fn):
    closed = []
    frontier = [(priority_fn(0, start), 0, start, [start])]
    expansions = 0

    while len(frontier) > 0:
        best_index = 0

        for i in range(len(frontier)):
            if frontier[i][0] < frontier[best_index][0]:
                best_index = i
        priority, g, node, path = frontier.pop(best_index)

        if node == goal:
            return path, g, expansions

        if node not in closed:
            closed.append(node)
            expansions += 1

            for neighbor, weight in graph[node]:
                new_g = g + weight
                new_path = path + [neighbor]

                frontier.append(
                    (priority_fn(new_g, neighbor),
                     new_g,
                     neighbor,
                     new_path)
                )

    return None, None, expansions

In [ ]:
path, cost, expansions = graph_search(
    sabcg_graph,
    'S',
    'G',
    astar_priority_sabcg
)
print("Fariha")
print("2254901100")
print("Path:", " -> ".join(path) if path else "No path found")
print("Total Cost:", cost)
print("Nodes Expanded:", expansions)

Fariha
2254901100
Path: S -> B -> C -> G
Total Cost: 6
Nodes Expanded: 4


Level 2: Real Dataset

In [ ]:
bd_graph_weighted = {
    'Rangpur': [('Bogura', 60), ('Mymensingh', 190)],
    'Bogura': [('Rangpur', 60), ('Rajshahi', 65), ('Dhaka', 210)],
    'Rajshahi': [('Bogura', 65), ('Jessore', 165)],
    'Jessore': [('Rajshahi', 165), ('Khulna', 55)],
    'Khulna': [('Jessore', 55), ('Barisal', 130), ('Dhaka', 210)],
    'Barisal': [('Khulna', 130), ('Dhaka', 165)],
    'Mymensingh': [('Rangpur', 190), ('Dhaka', 120)],
    'Dhaka': [('Bogura', 210), ('Khulna', 210), ('Barisal', 165),
              ('Mymensingh', 120), ('Comilla', 100)],
    'Comilla': [('Dhaka', 100), ('Chittagong', 150), ('Sylhet', 200)],
    'Sylhet': [('Comilla', 200)],
    'Chittagong': [('Comilla', 150), ("Cox's Bazar", 150)],
    "Cox's Bazar": [('Chittagong', 150)],
}

In [ ]:
bd_heuristics = {
    'Rangpur': 550,
    'Bogura': 480,
    'Rajshahi': 520,
    'Jessore': 470,
    'Khulna': 430,
    'Barisal': 330,
    'Mymensingh': 420,
    'Dhaka': 350,
    'Comilla': 220,
    'Sylhet': 350,
    'Chittagong': 130,
    "Cox's Bazar": 0,
}

In [ ]:
def astar_priority_bd(g, node):
    return g + bd_heuristics[node]

In [ ]:
def graph_search(graph, start, goal, priority_fn):
    closed = []
    frontier = [(priority_fn(0, start), 0, start, [start])]
    expansions = 0

    while len(frontier) > 0:
        best_index = 0

        for i in range(len(frontier)):
            if frontier[i][0] < frontier[best_index][0]:
                best_index = i

        priority, g, node, path = frontier.pop(best_index)

        if node == goal:
            return path, g, expansions

        if node not in closed:
            closed.append(node)
            expansions += 1

            for neighbor, weight in graph[node]:
                new_g = g + weight
                new_path = path + [neighbor]

                frontier.append(
                    (priority_fn(new_g, neighbor),
                     new_g,
                     neighbor,
                     new_path)
                )

    return None, None, expansions

Tasks 2: Search from Rangpur to Cox's Bazar.
1. Run A* and graph search. Report path, cost, and expansion count for each.

In [ ]:
path, cost, expansions = graph_search(
    bd_graph_weighted,
    'Rangpur',
    "Cox's Bazar",
    astar_priority_bd
)

print("Fariha")
print("2254901100")
print("Path:", " -> ".join(path) if path else "No path found")
print("Total Cost:", cost)
print("Nodes Expanded:", expansions)

Fariha
2254901100
Path: Rangpur -> Bogura -> Dhaka -> Comilla -> Chittagong -> Cox's Bazar
Total Cost: 670
Nodes Expanded: 7


Output Explanation:
A* finds the route Rangpur → Bogura → Dhaka → Comilla → Chittagong → Cox's Bazar. The total cost is 670, and A* expands 7 nodes before reaching the goal.

In [ ]:
def uniform_cost_priority(g, node):
    return g

In [ ]:
path, cost, expansions = graph_search(
    bd_graph_weighted,
    'Rangpur',
    "Cox's Bazar",
    uniform_cost_priority
)
print("Fariha")
print("2254901100")
print("Graph Search")
print("Path:", " -> ".join(path) if path else "No path found")
print("Total Cost:", cost)
print("Nodes Expanded:", expansions)

Fariha
2254901100
Graph Search
Path: Rangpur -> Bogura -> Dhaka -> Comilla -> Chittagong -> Cox's Bazar
Total Cost: 670
Nodes Expanded: 11


Output Explanation:
Graph search also finds the optimal path with a total cost of 670. However, it expands 11 nodes, which is more than A*(expands 7 nodes). Therefore, A* is more efficient for this dataset.

2. Trace the code for both search algorithms. Find the value of f (= g + h) for each step. Report if any edges violate admissibility and consistency.

In [ ]:
def trace_astar(graph, start, goal):
    closed = []
    frontier = [(bd_heuristics[start], 0, start, [start])]

    print("Node\tg\t h\t f")

    while len(frontier) > 0:
        best_index = 0

        for i in range(len(frontier)):
            if frontier[i][0] < frontier[best_index][0]:
                best_index = i

        f, g, node, path = frontier.pop(best_index)

        if node == goal:
            print(node, "\t", g, "\t", bd_heuristics[node], "\t", f)
            break

        if node not in closed:
            closed.append(node)

            print(node, "\t", g, "\t", bd_heuristics[node], "\t", f)

            for neighbor, weight in graph[node]:
                new_g = g + weight
                new_f = new_g + bd_heuristics[neighbor]

                frontier.append(
                    (new_f,
                     new_g,
                     neighbor,
                     path + [neighbor])
                )

In [ ]:
trace_astar(
    bd_graph_weighted,
    'Rangpur',
    "Cox's Bazar"
)

Node	g	 h	 f
Rangpur 	 0 	 550 	 550
Bogura 	 60 	 480 	 540
Mymensingh 	 190 	 420 	 610
Dhaka 	 270 	 350 	 620
Comilla 	 370 	 220 	 590
Rajshahi 	 125 	 520 	 645
Chittagong 	 520 	 130 	 650
Cox's Bazar 	 670 	 0 	 670


Trace Explanation:
A* selects the node with the lowest f-value, where f = g + h. It gradually moves from Rangpur toward Cox's Bazar and finally reaches the goal with a total cost of 670.

In [ ]:
def check_heuristics(graph, heuristics):
    print("Checking heuristic properties...\n")

    for node in graph:
        for neighbor, cost in graph[node]:
            admissible = heuristics[node] <= cost + heuristics[neighbor]
            consistent = heuristics[node] <= cost + heuristics[neighbor]

            if not admissible:
                print("Admissibility violation:",
                      node, "->", neighbor)

            if not consistent:
                print("Consistency violation:",
                      node, "->", neighbor)


In [ ]:
check_heuristics(bd_graph_weighted, bd_heuristics)

Checking heuristic properties...

Admissibility violation: Rangpur -> Bogura
Consistency violation: Rangpur -> Bogura
Admissibility violation: Dhaka -> Comilla
Consistency violation: Dhaka -> Comilla


 output explanation:
The heuristic is not fully admissible or consistent because some heuristic values are greater than the actual cost through neighboring nodes. Therefore, A* with this heuristic cannot be guaranteed to be optimal in general.